configtypes tests rely on regex cache

The tests broke in #962 when importing `cssutils`:

```
=============== FAILURES ===============
 TestRegex.test_transform_empty[foobar-expected0] 
self = <test_configtypes.TestRegex object at 0x7f47749b6630>
klass = <class 'qutebrowser.config.configtypes.Regex'>
val = 'foobar'
expected = re.compile('foobar')
    @pytest.mark.parametrize('val, expected', [
        (r'foobar', re.compile(r'foobar')),
        ('', None),
    ])
    def test_transform_empty(self, klass, val, expected):
>       assert klass().transform(val) == expected
E       assert re.compile('foobar') == re.compile('foobar')
E        +  where re.compile('foobar') = <bound method Regex.transform of <qutebrowser.config.configtypes.Regex object at 0x7f47803d1630>>('foobar')
E        +    where <bound method Regex.transform of <qutebrowser.config.configtypes.Regex object at 0x7f47803d1630>> = <qutebrowser.config.configtypes.Regex object at 0x7f47803d1630>.transform
E        +      where <qutebrowser.config.configtypes.Regex object at 0x7f47803d1630> = <class 'qutebrowser.config.configtypes.Regex'>()
tests/unit/config/test_configtypes.py:1044: AssertionError
 TestRegexList.test_transform[foo-expected0] 
self = <test_configtypes.TestRegexList object at 0x7f4774afd278>
klass = <class 'qutebrowser.config.configtypes.RegexList'>
val = 'foo'
expected = [re.compile('foo')]
    @pytest.mark.parametrize('val, expected', [
        ('foo', [re.compile('foo')]),
        ('foo,bar,baz', [re.compile('foo'), re.compile('bar'),
                         re.compile('baz')]),
        ('foo,,bar', [re.compile('foo'), None, re.compile('bar')]),
        ('', None),
    ])
    def test_transform(self, klass, val, expected):
>       assert klass().transform(val) == expected
E       assert [re.compile('foo')] == [re.compile('foo')]
E         At index 0 diff: re.compile('foo') != re.compile('foo')
E         Use -v to get the full diff
tests/unit/config/test_configtypes.py:1080: AssertionError
 TestRegexList.test_transform[foo,bar,baz-expected1] 
self = <test_configtypes.TestRegexList object at 0x7f4780566828>
klass = <class 'qutebrowser.config.configtypes.RegexList'>
val = 'foo,bar,baz'
expected = [re.compile('foo'), re.compile('bar'), re.compile('baz')]
    @pytest.mark.parametrize('val, expected', [
        ('foo', [re.compile('foo')]),
        ('foo,bar,baz', [re.compile('foo'), re.compile('bar'),
                         re.compile('baz')]),
        ('foo,,bar', [re.compile('foo'), None, re.compile('bar')]),
        ('', None),
    ])
    def test_transform(self, klass, val, expected):
>       assert klass().transform(val) == expected
E       assert [re.compile('...ompile('baz')] == [re.compile('f...ompile('baz')]
E         At index 0 diff: re.compile('foo') != re.compile('foo')
E         Use -v to get the full diff
tests/unit/config/test_configtypes.py:1080: AssertionError
 TestRegexList.test_transform[foo,,bar-expected2] 
self = <test_configtypes.TestRegexList object at 0x7f47747a6fd0>
klass = <class 'qutebrowser.config.configtypes.RegexList'>
val = 'foo,,bar'
expected = [re.compile('foo'), None, re.compile('bar')]
    @pytest.mark.parametrize('val, expected', [
        ('foo', [re.compile('foo')]),
        ('foo,bar,baz', [re.compile('foo'), re.compile('bar'),
                         re.compile('baz')]),
        ('foo,,bar', [re.compile('foo'), None, re.compile('bar')]),
        ('', None),
    ])
    def test_transform(self, klass, val, expected):
>       assert klass().transform(val) == expected
E       assert [re.compile('...ompile('bar')] == [re.compile('f...ompile('bar')]
E         At index 0 diff: re.compile('foo') != re.compile('foo')
E         Use -v to get the full diff
tests/unit/config/test_configtypes.py:1080: AssertionError
```

This is because the `re.compile` calls in the decorators happen at import time while the other ones happen later, and when importing `cssutils` in between, that causes the `re` cache to overflow... And since `re.compile(...).__eq__` checks for identity, those tests now fail...

Instead, we should probably have some `RegexEq` class to help compare regex objects:

```python
class RegexEq:

    def __init__(self, pattern, flags=0):
        pass

    def __eq__(self, other):
        if not isinstance(other, re.compile('')):
            return NotImplemented
        return (self.pattern, self.flags) == (other.pattern, other.flags)
```

and then use `RegexEq` instead of `re.compile` in the tests.

--- a/tests/unit/config/test_configtypes.py
+++ b/tests/unit/config/test_configtypes.py
@@ -75,7 +75,7 @@ class RegexEq:
 
     def __init__(self, pattern, flags=0):
         # We compile the regex because re.compile also adds flags defined in
-        # the pattern and implicit flags to it's .flags.
+        # the pattern and implicit flags to its .flags.
         # See https://docs.python.org/3/library/re.html#re.regex.flags
         compiled = re.compile(pattern, flags)
         self.pattern = compiled.pattern
@@ -1901,3 +1901,29 @@ class TestUserAgent:
     def test_complete(self, klass):
         """Simple smoke test for completion."""
         klass().complete()
+
+
+@pytest.mark.parametrize('first, second, equal', [
+    (re.compile('foo'), RegexEq('foo'), True),
+    (RegexEq('bar'), re.compile('bar'), True),
+    (RegexEq('qwer'), RegexEq('qwer'), True),
+    (re.compile('qux'), RegexEq('foo'), False),
+    (RegexEq('spam'), re.compile('eggs'), False),
+    (RegexEq('qwer'), RegexEq('rewq'), False),
+
+    (re.compile('foo', re.I), RegexEq('foo', re.I), True),
+    (RegexEq('bar', re.M), re.compile('bar', re.M), True),
+    (re.compile('qux', re.M), RegexEq('qux', re.I), False),
+    (RegexEq('spam', re.S), re.compile('eggs', re.S), False),
+
+    (re.compile('(?i)foo'), RegexEq('(?i)foo'), True),
+    (re.compile('(?i)bar'), RegexEq('bar'), False),
+])
+def test_regex_eq(first, second, equal):
+    if equal:
+        # Assert that the check is commutative
+        assert first == second
+        assert second == first
+    else:
+        assert first != second
+        assert second != first

--- a/tests/unit/config/test_configtypes.py
+++ b/tests/unit/config/test_configtypes.py
@@ -69,6 +69,34 @@ class NetworkProxy(QNetworkProxy):
                               password=self.password())
 
 
+class RegexEq:
+
+    """A class to compare regex objects."""
+
+    def __init__(self, pattern, flags=0):
+        # We compile the regex because re.compile also adds flags defined in
+        # the pattern and implicit flags to it's .flags.
+        # See https://docs.python.org/3/library/re.html#re.regex.flags
+        compiled = re.compile(pattern, flags)
+        self.pattern = compiled.pattern
+        self.flags = compiled.flags
+        self._user_flags = flags
+
+    def __eq__(self, other):
+        try:
+            # Works for RegexEq objects and re.compile objects
+            return (self.pattern, self.flags) == (other.pattern, other.flags)
+        except AttributeError:
+            return NotImplemented
+
+    def __repr__(self):
+        if self._user_flags:
+            return "RegexEq({!r}, flags={})".format(self.pattern,
+                                                    self._user_flags)
+        else:
+            return "RegexEq({!r})".format(self.pattern)
+
+
 @pytest.fixture
 def os_mock(mocker):
     """Fixture that mocks and returns os from the configtypes module."""
@@ -1037,7 +1065,7 @@ class TestRegex:
             klass().validate(val)
 
     @pytest.mark.parametrize('val, expected', [
-        (r'foobar', re.compile(r'foobar')),
+        (r'foobar', RegexEq(r'foobar')),
         ('', None),
     ])
     def test_transform_empty(self, klass, val, expected):
@@ -1070,10 +1098,10 @@ class TestRegexList:
             klass().validate(val)
 
     @pytest.mark.parametrize('val, expected', [
-        ('foo', [re.compile('foo')]),
-        ('foo,bar,baz', [re.compile('foo'), re.compile('bar'),
-                         re.compile('baz')]),
-        ('foo,,bar', [re.compile('foo'), None, re.compile('bar')]),
+        ('foo', [RegexEq('foo')]),
+        ('foo,bar,baz', [RegexEq('foo'), RegexEq('bar'),
+                         RegexEq('baz')]),
+        ('foo,,bar', [RegexEq('foo'), None, RegexEq('bar')]),
         ('', None),
     ])
     def test_transform(self, klass, val, expected):

